File: iOS 8 by Tutorials/Objective-C/Chapter 18 - Intermediate Scene Kit/PenguinPhysics/PenguinPhysics/GameViewController.m
//
//  GameViewController.m
//  PenguinPhysics
//
//  Created by Jake Gundersen on 7/25/14.
//  Copyright (c) 2014 Razeware, LLC. All rights reserved.
//

#import "GameViewController.h"

@interface GameViewController() <SCNSceneRendererDelegate>

@property (strong, nonatomic) IBOutlet SCNView *scnView;
@property (nonatomic, strong) SCNScene *scene;

@property (nonatomic, strong) SCNParticleSystem *snow;

@property (nonatomic, assign) BOOL applyForce;

@property (nonatomic, assign) float force;

@property (nonatomic, strong) SCNNode *cube;
@property (nonatomic, strong) SCNNode *rampNode;

@property (nonatomic, strong) SCNPhysicsSliderJoint *sliderJoint;

@end

@implementation GameViewController

- (void)viewDidLoad
{
  [super viewDidLoad];
  
  self.scene = [SCNScene sceneNamed:@"SceneIceRotated"];
  self.scnView.scene = self.scene;
  self.scnView.allowsCameraControl = YES;
  self.scnView.showsStatistics = YES;
  self.scnView.delegate = self;
  
  self.scene.physicsWorld.gravity = SCNVector3Make(0.0, -9.8, 0.0);
  
  self.cube = [self.scene.rootNode childNodeWithName:@"Cube" recursively:NO];
  self.cube.physicsBody = [SCNPhysicsBody dynamicBody];
  self.cube.physicsBody.mass = 5.0;
  self.cube.physicsBody.restitution = 0.01;
  
  SCNNode *ramp = [self.scene.rootNode childNodeWithName:@"Slope" recursively:NO];
  ramp.physicsBody = [SCNPhysicsBody kinematicBody];
  self.rampNode = [SCNNode node];
  [self.rampNode addChildNode:ramp];
  [self.scene.rootNode addChildNode:self.rampNode];
  self.rampNode.pivot = SCNMatrix4MakeTranslation(-7.0, 0.0, 0.0);
  self.rampNode.position = SCNVector3Make(-7.0, 0.0, 0.0);
  
  self.sliderJoint = [SCNPhysicsSliderJoint jointWithBodyA:self.cube.physicsBody axisA:SCNVector3Make(0.0, -1.0, 0.0) anchorA:SCNVector3Make(0.0, 0.0, -1.0) bodyB:ramp.physicsBody axisB:SCNVector3Make(0.0, -1.0, 0.0) anchorB:SCNVector3Make(0.0, 0.0, -0.20)];
  self.sliderJoint.maximumLinearLimit = 4.0;
  self.sliderJoint.minimumLinearLimit = -5.5;
  self.sliderJoint.minimumAngularLimit = 0.0;
  self.sliderJoint.maximumAngularLimit = 0.0;
  
  [self.scene.physicsWorld addBehavior:self.sliderJoint];
  
  self.cube.physicsBody.categoryBitMask = 1;
  self.cube.physicsBody.collisionBitMask = 1;
  ramp.physicsBody.categoryBitMask = 2;
  ramp.physicsBody.collisionBitMask = 2;
  
  SCNParticleSystem *snow = [SCNParticleSystem particleSystemNamed:@"Snow" inDirectory:nil];
  SCNNode *particleEmitterNode = [SCNNode node];
  particleEmitterNode.position = SCNVector3Make(0.0, 10.0, 0.0);
  [particleEmitterNode addParticleSystem:snow];
}

- (IBAction)applyForceTapped:(id)sender {
  UIButton *button = (UIButton *)sender;
  if (self.applyForce) {
    [button setTitle:@"Apply Force" forState:UIControlStateNormal];
    self.applyForce = NO;
  } else {
    [button setTitle:@"Stop" forState:UIControlStateNormal];
    self.applyForce = YES;
  }
}

- (IBAction)forceSliderMoved:(id)sender {
  UISlider *slider = (UISlider *)sender;
  self.force = slider.value * 300.0;
}

- (IBAction)angleSliderMoved:(id)sender {
  UISlider *slider = (UISlider *)sender;
  float angle = (1.0 - slider.value) * M_PI_4;
  self.rampNode.eulerAngles = SCNVector3Make(0.0, 0.0, angle);
}

- (void)renderer:(id<SCNSceneRenderer>)aRenderer updateAtTime:(NSTimeInterval)time {
  if (self.applyForce) {
    [self.cube.physicsBody applyForce:SCNVector3Make(self.force, 0.0, 0.0) impulse:NO];
  }
}


- (BOOL)shouldAutorotate
{
  return YES;
}

- (NSUInteger)supportedInterfaceOrientations
{
  if ([[UIDevice currentDevice] userInterfaceIdiom] == UIUserInterfaceIdiomPhone) {
    return UIInterfaceOrientationMaskAllButUpsideDown;
  } else {
    return UIInterfaceOrientationMaskAll;
  }
}

- (void)didReceiveMemoryWarning
{
  [super didReceiveMemoryWarning];
  // Release any cached data, images, etc that aren't in use.
}

@end

File: iOS 8 by Tutorials/Swift/Chapter 18 - Intermediate Scene Kit/PenguinPhysics-Final/PenguinPhysics/GameViewController.swift
//
//  GameViewController.swift
//  PenguinPhysics
//
//  Created by Jake Gundersen on 7/23/14.
//  Copyright (c) 2014 Razeware, LLC. All rights reserved.
//

import UIKit
import QuartzCore
import SceneKit

class GameViewController: UIViewController, SCNSceneRendererDelegate {
  
  @IBOutlet var scnView: SCNView!
  let scene = SCNScene(named: "SceneIceRotated.dae")!
  let snow = SCNParticleSystem(named: "Snow", inDirectory: nil)
  
  var applyForce = false
  var force:Float = 50.0
  
  let cube:SCNNode
  let slopeNode:SCNNode
  
  var sliderJoint:SCNPhysicsSliderJoint!
  
  required init(coder aDecoder: NSCoder) {
    cube = scene.rootNode.childNodeWithName("Cube", recursively: false)!
    slopeNode = SCNNode()
    
    super.init(coder: aDecoder)
  }
  
  override func viewDidLoad() {
    super.viewDidLoad()
    
    scnView.scene = scene
    scnView.allowsCameraControl = true
    scnView.showsStatistics = true
    scnView.delegate = self
    
    scene.physicsWorld.gravity = SCNVector3Make(0.0, -9.8, 0.0);
    
    cube.physicsBody = SCNPhysicsBody.dynamicBody()
    
    cube.physicsBody!.mass = 5.0
    cube.physicsBody!.restitution = 0.01
    
    if let slope = scene.rootNode.childNodeWithName("Slope", recursively: false) {
        slope.physicsBody = SCNPhysicsBody.kinematicBody()
        slopeNode.pivot = SCNMatrix4MakeTranslation(-7.0, 0.0, 0.0)
        slopeNode.position = SCNVector3Make(-7.0, 0.0, 0.0)
        slopeNode.addChildNode(slope)
        scene.rootNode.addChildNode(slopeNode)
        
        
        //1
        sliderJoint = SCNPhysicsSliderJoint(bodyA: cube.physicsBody, axisA: SCNVector3Make(0.0, -1.0, 0.0), anchorA: SCNVector3Make(0.0, 0.0, -1.0), bodyB: slope.physicsBody, axisB: SCNVector3Make(0.0, -1.0, 0.0), anchorB: SCNVector3Make(0.0, 0.0, -0.20))
        //2
        sliderJoint.maximumLinearLimit = 4.0
        sliderJoint.minimumLinearLimit = -5.5
        //3
        sliderJoint.maximumAngularLimit = 0.0
        sliderJoint.minimumAngularLimit = 0.0
        //4
        scene.physicsWorld.addBehavior(sliderJoint)
        
        cube.physicsBody!.categoryBitMask = 1
        cube.physicsBody!.collisionBitMask = 1
        slope.physicsBody!.categoryBitMask = 2
        slope.physicsBody!.collisionBitMask = 2
    }
    let particleEmitterNode = SCNNode()
    particleEmitterNode.addParticleSystem(snow)
    particleEmitterNode.position = SCNVector3(x: 0.0, y: 10.0, z: 0.0)
    scene.rootNode.addChildNode(particleEmitterNode)
  }
  
  override func shouldAutorotate() -> Bool {
    return true
  }
  
  @IBAction func applyForceTapped(sender: AnyObject) {
    let button = sender as UIButton
    if (applyForce) {
      button.setTitle("Apply Force", forState: .Normal)
      applyForce = false
    } else {
      button.setTitle("Stop", forState: .Normal)
      applyForce = true
    }
  }
  
  @IBAction func angleSliderMoved(sender: AnyObject) {
    var angle:Float = (1.0 - sender.value) * Float(M_PI_4)
    slopeNode.eulerAngles = SCNVector3Make(0.0, 0.0, angle)
  }
  
  @IBAction func forceSliderMoved(sender: AnyObject) {
    force = sender.value * 300.0
  }
  
  func renderer(aRenderer: SCNSceneRenderer, updateAtTime time: NSTimeInterval) {
    if applyForce {
      cube.physicsBody!.applyForce(SCNVector3Make(force, 0.0, 0.0), impulse: false)
    }
  }
  
  override func supportedInterfaceOrientations() -> Int {
      return Int(UIInterfaceOrientationMask.Landscape.rawValue)
  }
  
  override func didReceiveMemoryWarning() {
    super.didReceiveMemoryWarning()
    // Release any cached data, images, etc that aren't in use.
  }
  
}
